protoutils/protoutils.go

```diff
@@ -24,12 +26,45 @@ func InterfaceToMap(data interface{}) (map[string]interface{}, error) {
 	var err error
 	switch t.Kind() {
 	case reflect.Struct:
-		res, err = structToMap(data)
+		res, err = structToMap(data, true)
 		if err != nil {
 			return nil, err
 		}
 	case reflect.Map:
-		res, err = marshalMap(data)
+		res, err = marshalMap(data, true)
+		if err != nil {
+			return nil, err
+		}
+	case reflect.Array, reflect.Bool, reflect.Chan, reflect.Complex128, reflect.Complex64, reflect.Float32,
+		reflect.Float64, reflect.Func, reflect.Int, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Int8,
+		reflect.Interface, reflect.Invalid, reflect.Pointer, reflect.Slice, reflect.String, reflect.Uint,
+		reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uint8, reflect.Uintptr, reflect.UnsafePointer:
+		fallthrough
+	default:
+		return nil, errors.Errorf("data of type %T and kind %s not a struct or a map-like object", data, t.Kind().String())
+	}
+	return res, nil
+}
+
+//nolint:dupl
+func interfaceToMapKeepEmpty(data interface{}) (map[string]interface{}, error) {
+	if data == nil {
```

Comment:
To not duplicate the inner logic, could you have a interfaceToMap helper which takes `data` and `shouldOmitEmpty`? InterfaceToMap can call that with `true`, and then you can get rid of `interfaceToMapKeepEmpty` and just call the helper directly from `StructToStructPbIgnoreOmitEmpty`.